clicommand/build_cancel.go

```diff
@@ -0,0 +1,114 @@
+package clicommand
+
+import (
+	"context"
+	"fmt"
+
+	"time"
+
+	"github.com/buildkite/agent/v3/api"
+	"github.com/buildkite/agent/v3/logger"
+	"github.com/buildkite/roko"
+	"github.com/urfave/cli"
+)
+
+const buildCancelDescription = `Usage:
+
+    buildkite-agent build cancel [options...]
+
+Description:
+
+Cancel a running build.
+
+Example:
+
+    $ buildkite-agent build cancel --build "1234"`
+
+type BuildCancelConfig struct {
+	Build string `cli:"build" validate:"required"`
+
+	// Global flags
+	Debug       bool     `cli:"debug"`
+	LogLevel    string   `cli:"log-level"`
+	NoColor     bool     `cli:"no-color"`
+	Experiments []string `cli:"experiment" normalize:"list"`
+	Profile     string   `cli:"profile"`
+
+	// API config
+	DebugHTTP        bool   `cli:"debug-http"`
+	AgentAccessToken string `cli:"agent-access-token" validate:"required"`
+	Endpoint         string `cli:"endpoint" validate:"required"`
+	NoHTTP2          bool   `cli:"no-http2"`
+}
+
+var BuildCancelCommand = cli.Command{
+	Name:        "cancel",
+	Usage:       "Cancel a build",
+	Description: buildCancelDescription,
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:   "build",
+			Value:  "",
+			Usage:  "The build to cancel",
+			EnvVar: "BUILDKITE_BUILD_ID",
+		},
+
+		// API Flags
+		AgentAccessTokenFlag,
+		EndpointFlag,
+		NoHTTP2Flag,
+		DebugHTTPFlag,
+
+		// Global flags
+		NoColorFlag,
+		DebugFlag,
+		LogLevelFlag,
+		ExperimentsFlag,
+		ProfileFlag,
+	},
+	Action: func(c *cli.Context) error {
+		ctx := context.Background()
+		ctx, cfg, l, _, done := setupLoggerAndConfig[BuildCancelConfig](ctx, c)
+		defer done()
+
+		if err := cancelBuild(ctx, cfg, l); err != nil {
```

Comment:
Well that's just

```suggestion
		return cancelBuild(ctx, cfg, l)
```

You could consider lifting the whole action out of the struct literal, e.g.

```go
    Action: cancelBuildAction,
}

func cancelBuildAction(c *cli.Context) error {
   ctx := context.Background()
   ctx, cfg, l, _, done := setupLoggerAndConfig[BuildCancelConfig](ctx, c)
   defer done()
   
   // body of cancelBuild here
}
```

Reply:
> You could consider lifting the whole action out of the struct literal, e.g.

I had a quick play but couldn't find a nice way to inject test config into the cli context. I think for now I'll leave them as two seperate functions if you're alright with that?